### PR TITLE
internal/witcli: auto-create output directory if it does not exist

### PR DESCRIPTION
**Pull request by joonas**
Having felt the need for #213 a bunch of times, I figured it would be nice to implement this.

For context I did connect with @brooksmtownsend to confirm he was good with me implementing this before jumping on it.

### CODE REVIEW

**Comment by joonas** [comment]
@ydnar lmk if the tests I added don't line up with what you had in mind, I figured it was worth testing both Find and Create side of the functionality here.

File: internal/witcli/witcli_test.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
+package witcli_test
```
Comment (ydnar) [inline internal/witcli/witcli_test.go:1]: This module doesn't use test packages, is there a specific reason?

**Review by ydnar — APPROVED** [review]
LGTM except the test package name. Thank you!

File: internal/witcli/witcli_test.go (hunk 1 above)
Comment (joonas) [reply internal/witcli/witcli_test.go]: Oversight on my part, fixed!

File: internal/witcli/witcli_other.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
+//go:build !tinygo && !tinygo.wasm && !wasip1 && !wasip2
```
Comment (ydnar) [inline internal/witcli/witcli_other.go:1]: I have a fix for this that removes the need for build tags on this.

**Comment by ydnar** [comment]
Once #218 is merged, please check out the changes in #217 and apply them here.

I’d like to avoid cascading build tag complexity when the root cause is dependencies of package `regclient` like `logrus` and `libtrust`.

**Comment by ydnar** [comment]
Also: this projects follows Go project convention of `package: present-tense description of change`: https://go.dev/wiki/CommitMessage

TODO for me: update the README!

**Comment by joonas** [comment]
@ydnar thank you for the feedback, I'll walk through it and make the changes here!

**Comment by joonas** [comment]
@ydnar appreciate your feedback and patience here, the PR should be in a good spot now

File: internal/witcli/dir.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+	"os"
+)
+
+// FindOrCreateDir attempts to load file information for the provided path and
+// the provided path does not already exist, it attempts to create the directory
+// and then return the file information for it.
+func FindOrCreateDir(path string) (fs.FileInfo, error) {
+	info, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		err = os.MkdirAll(path, os.FileMode(os.ModeDir|os.ModePerm))
```
Comment (Mossaka) [inline internal/witcli/dir.go:15]: IIUC, `os.ModePerm` is 0x777 that gives full permissions for the owner, group, and others. Do we really want it? Perhaps we only want `0755 `

File: internal/witcli/dir.go [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+	"io/fs"
+	"os"
+)
+
+// FindOrCreateDir attempts to load file information for the provided path and
+// the provided path does not already exist, it attempts to create the directory
+// and then return the file information for it.
+func FindOrCreateDir(path string) (fs.FileInfo, error) {
+	info, err := os.Stat(path)
+	if os.IsNotExist(err) {
```
Comment (Mossaka) [inline internal/witcli/dir.go:14]: **suggest**: use `if errors.Is(err, fs.ErrNotExist)`

File: internal/witcli/dir_test.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
+package witcli
```
Comment (Mossaka) [inline internal/witcli/dir_test.go:1]: would be nice to add a test case where it tries to create a diretory where permissions are restricted. 

E.g.

```Go

func TestFindOrCreateDirPermissionDenied(t *testing.T) {
	if runtime.GOOS == "windows" {
	    return
	}
	_, err := FindOrCreateDir("/root/protected")
	if err == nil || !os.IsPermission(err) {
		t.Errorf("Expected permission error, got: %v", err)
	}
}
```

File: internal/witcli/dir.go (hunk 3 above)
Comment (ydnar) [reply internal/witcli/dir.go:15]: Agree

**Review by ydnar — COMMENTED** [review]
Thanks!

File: internal/witcli/dir.go (hunk 3 above)
Comment (joonas) [reply internal/witcli/dir.go:15]: I'd based this off of the [original comment here](https://github.com/bytecodealliance/wasm-tools-go/issues/213#issuecomment-2436257272), sounds like we're good with `0o755` instead?

If so, I'm happy to make the change, just wanted to confirm

File: internal/witcli/dir.go (hunk 4 above)
Comment (joonas) [reply internal/witcli/dir.go]: Good call, I'd honestly not realized that this wasn't a direct translation, but the docs state as much, thanks!

File: internal/witcli/dir_test.go (hunk 5 above)
Comment (joonas) [reply internal/witcli/dir_test.go:1]: I like that suggestion, added one 👍

**Review by Mossaka — APPROVED** [review]

**Review by ydnar — APPROVED** [review]

**Pull request merged**